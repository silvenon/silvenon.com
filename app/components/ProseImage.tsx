import type CSS from 'csstype'
import cloudinary from '../utils/cloudinary'
import clsx from 'clsx'

// approximate px values of @tailwindcss/typography's 65ch on different viewports
const proseMaxWidth = {
  sm: '620px',
  lg: '700px',
  xl: '780px',
  '2xl': '920px',
}
const imageSizes = [640, 768, 1024, 1536, 2048]
const largestImageSize = Math.max(...imageSizes)

export function getImageUrl(id: string, size: number) {
  return cloudinary(id, {
    crop: 'limit',
    width: size,
    format: 'auto',
    quality: 'auto',
  })
}

interface Props extends React.ImgHTMLAttributes<HTMLImageElement> {
  width: number
  height: number
  cloudinaryId: string
}

export default function ProseImage({
  width,
  height,
  cloudinaryId,
  ...props
}: Props) {
  const maxWidth = width < largestImageSize ? width / 2 : undefined

  return (
    <div className="not-prose mx-auto" style={{ maxWidth }}>
      {/* biome-ignore lint/a11y/useAltText: alt is passed through props */}
      <img
        {...props}
        className={clsx(
          props.className,
          'aspect-[var(--width)/var(--height)] w-full rounded-lg',
        )}
        style={{
          '--width': width,
          '--height': height,
        }}
        loading="lazy"
        src={getImageUrl(cloudinaryId, 1024)}
        srcSet={imageSizes
          .filter((size) => size < width)
          .concat(width)
          .map((size) => `${getImageUrl(cloudinaryId, size)} ${size}w`)
          .join(', ')}
        sizes={[
          `(min-width: ${import.meta.env.SCREEN_2XL}) ${proseMaxWidth['2xl']}`,
          `(min-width: ${import.meta.env.SCREEN_XL}) ${proseMaxWidth.xl}`,
          `(min-width: ${import.meta.env.SCREEN_LG}) ${proseMaxWidth.lg}`,
          `(min-width: ${import.meta.env.SCREEN_SM}) ${proseMaxWidth.sm}`,
          `calc(100vw - (var(--page-padding) * 2))`,
        ].join(', ')}
      />
    </div>
  )
}

declare module 'react' {
  interface CSSProperties extends CSS.Properties<string | number> {
    '--width'?: number
    '--height'?: number
  }
}
