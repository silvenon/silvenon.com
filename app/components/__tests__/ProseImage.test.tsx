import { render } from '@testing-library/react'
import ProseImage from '../ProseImage'

describe('ProseImage', () => {
  it('has correct output', () => {
    const { container } = render(
      <ProseImage
        alt="description"
        cloudinaryId="1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl"
        width={800}
        height={600}
      />,
    )

    // eslint-disable-next-line testing-library/no-node-access
    expect(container.firstChild).toMatchSnapshot()
  })
})
