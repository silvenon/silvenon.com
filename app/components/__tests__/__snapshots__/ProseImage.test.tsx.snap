// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProseImage > has correct output 1`] = `
<div
  class="not-prose mx-auto"
  style="max-width: 400px;"
>
  <img
    alt="description"
    class="aspect-[var(--width)/var(--height)] w-full rounded-lg"
    loading="lazy"
    sizes="(min-width: 96rem) 920px, (min-width: 80rem) 780px, (min-width: 64rem) 700px, (min-width: 40rem) 620px, calc(100vw - (var(--spacing-page) * 2))"
    src="https://res.cloudinary.com/silvenon/image/upload/c_limit,w_1024,f_auto,q_auto/1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl"
    srcset="https://res.cloudinary.com/silvenon/image/upload/c_limit,w_640,f_auto,q_auto/1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl 640w, https://res.cloudinary.com/silvenon/image/upload/c_limit,w_768,f_auto,q_auto/1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl 768w, https://res.cloudinary.com/silvenon/image/upload/c_limit,w_800,f_auto,q_auto/1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl 800w"
    style="--width: 800; --height: 600;"
  />
</div>
`;
